src/ray/core_worker/actor_handle.cc

```diff
@@ -58,10 +58,25 @@ void ActorHandle::SetActorTaskSpec(TaskSpecBuilder &builder,
 
 void ActorHandle::Serialize(std::string *output) { inner_.SerializeToString(output); }
 
-void ActorHandle::Reset() {
-  std::unique_lock<std::mutex> guard(mutex_);
-  task_counter_ = 0;
-  actor_cursor_ = ObjectID::FromBinary(inner_.actor_cursor());
+void ActorHandle::Connect(const gcs::ActorTableData &data,
+                          rpc::ClientCallManager &client_call_manager) {
+  RAY_CHECK(data.state() == gcs::ActorTableData::ALIVE);
+  data_.reset(new gcs::ActorTableData(data));
+  rpc_client_ = rpc::DirectActorClient::make(data_->ip_address(), data_->port(),
+                                             client_call_manager);
+}
+
+void ActorHandle::Reset(const gcs::ActorTableData &data, bool reset_task_counter) {
+  RAY_CHECK(data.state() != gcs::ActorTableData::ALIVE);
+  data_.reset(new gcs::ActorTableData(data));
+  // Disconnect from the RPC client.
+  rpc_client_.reset();
+
+  if (reset_task_counter) {
```

Comment:
Why not always reset?

Reply:
```
        // We have to reset the actor handle since the next instance of the
        // actor will not have the last sequence number that we sent.
        // TODO: Remove the flag for direct calls. We do not reset for the
        // raylet codepath because it tries to replay all tasks since the last
        // actor checkpoint.
```

Reply:
Ah, the mythical task replay.